src/login/oidc/refresh/TokenRefresher.ts

```diff
@@ -0,0 +1,120 @@
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { injectable, inject } from "tsyringe";
+import { IFetcher } from "../../../util/Fetcher";
+import { IStorageUtility } from "../../../localStorage/StorageUtility";
+import { IIssuerConfigFetcher } from "../IssuerConfigFetcher";
+import { IDpopHeaderCreator } from "../../../dpop/DpopHeaderCreator";
+import URL from "url-parse";
+import formurlencoded from "form-urlencoded";
+import { ITokenSaver } from "../redirectHandler/TokenSaver";
+
+export interface ITokenRefresher {
+  refresh(localUserId: string): Promise<void>;
+}
+
+@injectable()
+export default class TokenRefresher implements ITokenRefresher {
+  constructor(
+    @inject("fetcher") private fetcher: IFetcher,
+    @inject("issuerConfigFetcher")
+    private issuerConfigFetcher: IIssuerConfigFetcher,
+    @inject("storageUtility") private storageUtility: IStorageUtility,
+    @inject("dpopHeaderCreator") private dpopHeaderCreator: IDpopHeaderCreator
+  ) {}
+
+  async refresh(localUserId: string): Promise<void> {
+    // Get the refresh token and the issuer
+    const [refreshToken, issuer, clientId, clientSecret] = await Promise.all([
+      this.storageUtility.getForUser(localUserId, "refreshToken", true),
+      this.storageUtility.getForUser(localUserId, "issuer", true),
+      this.storageUtility.getForUser(localUserId, "clientId", true),
+      this.storageUtility.getForUser(localUserId, "clientSecret")
+    ]);
+
+    // Get the issuer config to find the token url
+    const issuerConfig = await this.issuerConfigFetcher.fetchConfig(
+      new URL(issuer as string)
+    );
+    if (
+      !issuerConfig.grantTypesSupported ||
+      !issuerConfig.grantTypesSupported.includes("refresh_token")
+    ) {
+      throw new Error(
+        `The issuer ${issuer} does not support the refresh token grant`
+      );
+    }
+    if (!issuerConfig.tokenEndpoint) {
+      throw new Error(`This issuer ${issuer} does not have a token endpoint`);
+    }
+
+    // Perform refresh token flow
+    const tokenRequestInit: RequestInit = {
+      method: "POST",
+      headers: {
+        DPoP: await this.dpopHeaderCreator.createHeaderToken(
+          issuerConfig.tokenEndpoint,
+          "POST"
+        ),
+        "content-type": "application/x-www-form-urlencoded"
+      },
+      /* eslint-disable @typescript-eslint/camelcase */
+      body: formurlencoded({
+        client_id: clientId as string,
+        grant_type: "refresh_token",
+        refresh_token: refreshToken
+      })
+      /* eslint-enable @typescript-eslint/camelcase */
+    };
+
+    if (clientSecret) {
+      (tokenRequestInit.headers as Record<
+        string,
+        string
+      >).Authorization = `Basic ${this.btoa(`${clientId}:${clientSecret}`)}`;
+    }
+
+    const tokenResponse = await (
```

Comment:
What if the response is invalid? You shouldn't need the casts below (`as string`); ideally, a validator acts as a type guard.